File: go/2015/day01/main.go
package main

import (
	"fmt"
	"os"
)

func part1(input string) int {
	floor := 0

	for _, c := range input {
		if c == '(' {
			floor += 1
		} else if c == ')' {
			floor -= 1
		}
	}
	return floor
}

func part2(input string) int {
	floor := 0
	for i, c := range input {
		if c == '(' {
			floor += 1
		} else if c == ')' {
			floor -= 1
		}
		if floor < 0 {
			return i + 1
		}
	}
	return 0
}

func main() {
	b, err := os.ReadFile("input.txt")

	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	str := string(b)

	fmt.Println(part1(str))
	fmt.Println(part2(str))
}

File: python/2015/day01/main.py
from collections import Counter


def pt1(input: str) -> int:
    c = Counter(input)
    return c["("] - c[")"]


def pt2(input: str) -> int:
    floor = 0
    for i, c in enumerate(input):
        match c:
            case "(":
                floor = floor + 1
            case ")":
                floor = floor - 1
        if floor < 0:
            return i + 1
    return -1


if __name__ == "__main__":
    with open("day01/input.txt", "r") as fin:
        input = fin.read().strip()
        print(pt1(input))
        print(pt2(input))
